src/omv/imlib/jpeg_dec.c

```diff
@@ -23,13 +23,3400 @@
 #include "imlib.h"
 #include "omv_boardconfig.h"
 
-#define TIME_JPEG   (0)
+#include "JPEGDEC.h"
+
+#define TIME_JPEG   (1)
 #if (TIME_JPEG == 1)
 #include "py/mphal.h"
 #endif
 
-// API OUTDATED!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
 //
+// forward references
+static int JPEGInit(JPEGIMAGE *pJPEG);
+static int JPEGParseInfo(JPEGIMAGE *pPage, int bExtractThumb);
+static void JPEGGetMoreData(JPEGIMAGE *pPage);
+static int DecodeJPEG(JPEGIMAGE *pImage);
+static int32_t readRAM(JPEGFILE *pFile, uint8_t *pBuf, int32_t iLen);
+static int32_t seekMem(JPEGFILE *pFile, int32_t iPosition);
+static void JPEGDither(JPEGIMAGE *pJPEG, int iWidth, int iHeight);
+/* JPEG tables */
+// zigzag ordering of DCT coefficients
+static const unsigned char cZigZag[64] = {0,1,5,6,14,15,27,28,
+    2,4,7,13,16,26,29,42,
+    3,8,12,17,25,30,41,43,
+    9,11,18,24,31,40,44,53,
+    10,19,23,32,39,45,52,54,
+    20,22,33,38,46,51,55,60,
+    21,34,37,47,50,56,59,61,
+    35,36,48,49,57,58,62,63};
+
+// un-zigzag ordering
+static const unsigned char cZigZag2[64] = {0,1,8,16,9,2,3,10,
+    17,24,32,25,18,11,4,5,
+    12,19,26,33,40,48,41,34,
+    27,20,13,6,7,14,21,28,
+    35,42,49,56,57,50,43,36,
+    29,22,15,23,30,37,44,51,
+    58,59,52,45,38,31,39,46,
+    53,60,61,54,47,55,62,63};
+
+// For AA&N IDCT method, multipliers are equal to quantization
+// coefficients scaled by scalefactor[row]*scalefactor[col], where
+// scalefactor[0] = 1
+// scalefactor[k] = cos(k*PI/16) * sqrt(2)    for k=1..7
+// For integer operation, the multiplier table is to be scaled by
+// IFAST_SCALE_BITS.
+static const int iScaleBits[64] = {16384, 22725, 21407, 19266, 16384, 12873,  8867,  4520,
+    22725, 31521, 29692, 26722, 22725, 17855, 12299,  6270,
+    21407, 29692, 27969, 25172, 21407, 16819, 11585,  5906,
+    19266, 26722, 25172, 22654, 19266, 15137, 10426,  5315,
+    16384, 22725, 21407, 19266, 16384, 12873,  8867,  4520,
+    12873, 17855, 16819, 15137, 12873, 10114,  6967,  3552,
+    8867, 12299, 11585, 10426,  8867,  6967,  4799,  2446,
+    4520,  6270,  5906,  5315,  4520,  3552,  2446,  1247};
+//
+// Range clip and shift for RGB565 output
+// input value is 0 to 255, then another 256 for overflow to FF, then 512 more for negative values wrapping around
+// Trims a few instructions off the final output stage
+//
+static const uint8_t ucRangeTable[] = {0x80,0x81,0x82,0x83,0x84,0x85,0x86,0x87,0x88,0x89,0x8a,0x8b,0x8c,0x8d,0x8e,0x8f,
```

Comment:
ARM has an instruction that does a shift and clips in one cycle. Please use it versus a range table.